sky/backends/provision_utils.py

```diff
@@ -0,0 +1,415 @@
+    """Deleting or stopping a cluster."""
+    if terminate:
+        provision.terminate_instances(cloud_name, region, cluster_name)
+        metadata_utils.remove_cluster_metadata(cluster_name)
+    else:
+        provision.stop_instances(cloud_name, region, cluster_name)
+
+    # TODO(suquark): In theory, users do not need to wait for the cluster
+    #  to be taken down completely and can interrupt the CLI at any time.
+    #  However, we cannot notify users this fact because there is a
+    #  progress bar currently running.
+    status_to_wait = 'terminated' if terminate else 'stopped'
+    provision.wait_instances(cloud_name, region, cluster_name, status_to_wait)
+
+
+def _wait_ssh_connection_direct(ip: str) -> bool:
+    try:
+        with socket.create_connection((ip, 22), timeout=1) as s:
+            if s.recv(100).startswith(b'SSH'):
+                return True
+    except socket.timeout:  # this is the most expected exception
+        pass
+    except Exception:  # pylint: disable=broad-except
+        pass
+    return False
+
+
+def _wait_ssh_connection_indirect(
+        ip: str,
+        ssh_user: str,
+        ssh_private_key: str,
+        ssh_control_name: Optional[str] = None,
+        ssh_proxy_command: Optional[str] = None) -> bool:
+    del ssh_control_name
+    # We test ssh with 'echo', because it is of the most common
+    # commandline programs on both Unix-like and Windows platforms.
+    # NOTE: Ray uses 'uptime' command and 10s timeout.
+    command = [
+        'ssh', '-T', '-i', ssh_private_key, f'{ssh_user}@{ip}', '-o',
+        'StrictHostKeyChecking=no', '-o', 'ConnectTimeout=20s', '-o',
+        f'ProxyCommand={ssh_proxy_command}', 'echo'
+    ]
+    proc = subprocess.run(command,
+                          shell=False,
+                          check=False,
+                          stdout=subprocess.DEVNULL,
+                          stderr=subprocess.DEVNULL)
+    return proc.returncode == 0
+
+
+def wait_for_ssh(cluster_metadata: provision_comm.ClusterMetadata,
+                 ssh_credentials: Dict[str, str]):
+    """Wait until SSH is ready."""
+    ips = cluster_metadata.get_feasible_ips()
+    if cluster_metadata.has_public_ips():
+        # If we can access public IPs, then it is more efficient to test SSH
+        # connection with raw sockets.
+        waiter = _wait_ssh_connection_direct
+    else:
+        # See https://github.com/skypilot-org/skypilot/pull/1512
+        waiter = functools.partial(_wait_ssh_connection_indirect,
+                                   **ssh_credentials)
+
+    timeout = 60 * 10  # 10-min maximum timeout
+    start = time.time()
+    # use a queue for SSH querying
+    ips = collections.deque(ips)
+    while ips:
+        ip = ips.popleft()
+        if not waiter(ip):
+            ips.append(ip)
+            if time.time() - start > timeout:
+                with ux_utils.print_exception_no_traceback():
+                    raise TimeoutError(
```

Comment:
Consider raising RuntimeError -- such an error means something's wrong with subnet and it's not easily recoverable by retrying.